flixel/math/FlxVector.hx

```diff
@@ -6,6 +6,149 @@ import openfl.geom.Point;
+ * // and assign it to new variable C
+ * var C = FlxVector.get(2, 2);
+ * 
+ * // B and C points to the same object in memory
+ * B.negate();
+ * 
+ * trace(C);	// (2, 2) expected, but it is (-2, -2)
+ * ```
+ * 
+ * This problem with `B` and `C` above is not specific to overloaded operators,
+ * you can fall into it using regular methods too.
+ * As of yet (4.10.0) Flixel does not warn you in any way if you use disposed vectors.
+ * In general, you should know what kind of vectors you are using and handle 
+ * them properly. You can also use non-weak vectors and dispose them manually
+ * when you are done with them:
+ * 
+ * ```haxe
+ * // Example 2
+ * 
+ * var A = FlxVector.get(1, 2);
+ * var B = FlxVector.get(3, 4);
+ * 
+ * A -= B;
+ * 
+ * var C = FlxVector.get(2, 2);
+ * B.negate();
+ * trace(C);	// (2, 2) - that is correct
+ * 
+ * A.put();		// utilize your vectors when they aren't needed anymore
+ * B.put();
+ * C.put();
+ * ```
+ * 
+ * ### Binary and unary operators
+ * 
+ * While operators from the first group is for modifying existing vectors, 
+ * second group operators are meant to return new vectors and *similar* to the methods
+ * like `addPointNew()`. Similar, but **not** the same.
+ * 
+ * - `A + B` returns a new `weak` vector that is sum of `A` and `B`
+ * - `A - B` returns a new `weak` vector that is difference of `A` and `B`
+ * - `A * k` returns a new `weak` vector that is `A` scaled with coefficient `k`
+ * - `-A` returns a new `weak` vector that is negated version of `A`
+ * - `A * B` returns dot product of `A` and `B`
+ * 
+ * Two peculiarities of these operators:
+ * 
+ * 1. They return new `weak` vector (except for dot product, its result cannot be
+ *    other than `Float`)
+ * 2. Any `weak` vector operand will be released to the pool
+ * 
+ * Charged with these features, it is possible to build complex expressions
+ * without worrying that intermediate results lost in memory waiting to be
+ * garbage-collected.
+ * 
+ * ```haxe
+ * // Example 3
+ * 
+ * var A = FlxVector.weak(1, 2);
+ * var B = FlxVector.weak(3, 4);
+ * 
+ * var D = -(A + B);
+ * 
+ * trace(D);	// (-4, -6)
+ * ```
+ * 
+ * Expression `-(A + B)` can be rewritten like this:
+ * 
+ * ```haxe
+ * // Example 4
+ * 
+ * var C = A + B;	// both A and B are weak, so they both return to the pool
+ * 
+ * // C is weak, it returns to the pool in negate operation:
+ * var D = -C;
+ * 
+ * // NOTE: D is weak too, since it is the result of `-C` negate operation
+ * ```
+ * 
+ * If you want to use a weak result of an expression more than once,
+ * you need to "unweak" it:
+ * 
+ * ```haxe
+ * // Example 5
+ * 
+ * var A = FlxVector.weak(1, 1);
+ * var B = FlxVector.weak(2, 2);
+ * 
+ * var C = (A + B);		// A and B return to pool, C is weak
+ * 
+ * // "unweak" vector C:
+ * // We create a new non weak vector and copy vector C into it. 
+ * // Original vector C returns to pool in `copyFrom()`
+ * C = FlxVector.get().copyFrom(C);
```

Comment:
As I said before, even if it we clearly documented this, people still will have problems, and not know why. automatic weak vectors are unintuitive and problematic. I'd much rather have GC'd instances from intermediary vectors as a trade-off for the convenience of operators.

The whole point of these operators is convenience and simpler code and this implementations is not intuitive, convenient or simple.

Reply:
The way that "problematic" operators implemented (they are hidden from completion) dictate the way how users find them (they thoroughly read `FlxVector` docs). Thus it should not be a concern if someone misuses them.  And again, flixel must prohibit using disposed vectors in a loud and explicit way, so the users of "problematic" operators get such error as soon as possible.

>this implementation is not intuitive, convenient or simple

Math operations on complex objects just cannot be intuitive, convenient or simple. All these three terms might be understood differently by different people.

Suggestions:

- type of the result vector can be controlled with some global flag on `FlxVector` class
  ```haxe
  public static inline var EXPRESSIONS_RETURN_WEAK = true;

  static inline function sumVectorsOp(lhs:FlxVector, rhs:FlxVector):FlxVector
  {
    var result = weak(lhs.x, lhs.y).addPoint(rhs);
    lhs.putWeak();
    result._weak = EXPRESSIONS_RETURN_WEAK;
    return result;	
  }
- unweak vectors with macro (if this is possible to write such macro)

Reply:
> The way that "problematic" operators implemented (they are hidden from completion) dictate the way how users find them (they thoroughly read FlxVector docs). Thus it should not be a concern if someone misuses them.

As someone who routinely aids newcomers to flixel, can 100% say that people do not read the doc. Most of them learn from reading or copying from other people code. Its quite common that i have to explain the basic duty of a function they found but did not read into. Plus anyone coming from other languages with vector operators will likely just try them out or assume they exist.

Reply:
Apparently we may be able to avoid this argument entirely by inline the constuctors of FlxPoint and FlxVector, and that it might even work without having to remove our existing pooling system. This seems like something worth looking into

Reply:
> inline the constuctors of FlxPoint and FlxVector

Doesn't it require complete reworking of  `FlxPoint` ? Which affects `FlxCallbackPoint` and so on.